[Xinbei] iP

xinbei-shen opened the pull request:
### Duke 🐧
---
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use
- pleasant to interact with
- **FREE!!**

All you need to do is, 

1. Download it [here](https://github.com/ichigh0st/ip/releases/tag/A-Jar)
2. double-click it
3. add your tasks
4. and let it remember your tasks for you!

Unique features:

- [X] Managing tasks
- [x] Managing deadlines (coming soon!)
- [ ] Reminders (coming soon!)
---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Duke {
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

xinbei-shen committed:
Implement skeletal version of Duke

xinbei-shen committed:
Save and display inputs in a list

xinbei-shen committed:
Mark tasks as done and undone

xinbei-shen committed:
Handle 3 types of tasks

xinbei-shen committed:
Handle errors

xinbei-shen committed:
Delete tasks

xinbei-shen committed:
Semi-automate testing of Duke

xinbei-shen committed:
Encapsulate the task list in its own class

xinbei-shen committed:
Save task list to the hard disk

This allows the information previously entered to be
loaded upon starting up Duke again.

xinbei-shen committed:
Add support for processing date and times

xinbei-shen committed:
Merge branch 'branch-Level-7'

xinbei-shen committed:
Merge branch 'branch-Level-8'

xinbei-shen committed:
Add further OOP and organize classes into Packages

xinbei-shen committed:
Add gradle support

xinbei-shen committed:
Add JUnit tests

xinbei-shen committed:
Add JavaDoc comments

xinbei-shen committed:
Update .gitignore

xinbei-shen committed:
Follow coding standard

xinbei-shen committed:
Find task using keywords

xinbei-shen committed:
Merge branch 'branch-A-JavaDoc'

xinbei-shen committed:
Merge branch 'branch-A-CodingStandard'

xinbei-shen committed:
Merge branch 'branch-Level-9'

kangqiao322 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,100 @@
+     * @param searchDate The date to be checked.
+     * @return True if the task occurs on the given date.
+     */
+    public boolean happensOn(LocalDate searchDate) {
```
Perhaps it would be better to name Boolean variable with a prefix such as is?

xinbei-shen replied:
you're right thanks!

kangqiao322 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,76 @@
+ *
+ * Duke will stop running upon receiving the "bye" command.
+ */
+public class Duke {
+    private final Storage storage;
+    private final Ui ui;
```
I think it would be better for constants to have names that are all uppercase.

kangqiao322 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,127 @@
+        String[] arr = input.split(" ", 2);
+        String command = arr[0];
+
+        switch (command) {
+        case ByeCommand.COMMAND_WORD:
+            return new ByeCommand();
+        case ListCommand.COMMAND_WORD:
+            return new ListCommand();
+        case MarkCommand.COMMAND_WORD:
+            try {
+                return new MarkCommand(Integer.parseInt(arr[1]) - 1);
+            } catch (NumberFormatException | IndexOutOfBoundsException e) {
+                ui.invalidMarkInput();
+            }
+            return null;
+        case UnmarkCommand.COMMAND_WORD:
+            try {
+                return new UnmarkCommand(Integer.parseInt(arr[1]) - 1);
+            } catch (NumberFormatException | IndexOutOfBoundsException e) {
+                ui.invalidUnmarkInput();
+            }
```
I like that you use switch cases for numerous conditions, it looks much cleaner. Great work!

kangqiao322 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,127 @@
+            return null;
+        case EventCommand.COMMAND_WORD:
+            try {
+                String[] info = arr[1].split(" /", 2);
+                String[] timings = info[1].split(" ", 2);
+                String[] dateTimeInfo = timings[1].split(" - ");
+                LocalDateTime startDateTime = LocalDateTime.parse(dateTimeInfo[0], dateTimeFormatter);
```
I like that you kept the same level of abstraction in each code fragment.👍

kangqiao322 reviewed (commented):
Overall a work well done! Each method also has its respective JavaDocs as well which makes it much easier to understand its functionalities. However, some names may need to be changed.

sjoann commented on src/test/java/duke/task/TaskListTest.java:
```diff
@@ -0,0 +1,50 @@
+                output.toString());
+    }
+
+    @Test
```
I like that you created multiple tests for one class

sjoann commented on src/main/java/duke/task/Task.java:
```diff
@@ -12,23 +15,44 @@ public class Task {
+     * TaskType values are used to indicate the type of the Task
+     * to retrieve the corresponding information.
+     */
     public enum TaskType {
```
I like that you created enums for this. However, I think enum names must be written in PascalCase.

xinbei-shen replied:
oops you're right thanks!

sjoann commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -23,11 +23,22 @@
 public class Parser {
     private static DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
     private static DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HHmm");
     private static DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
```
Good effort in having different date formatting :)

sjoann reviewed (commented):
I like that your code is easy to read and good effort in creating the JavaDocs :)

xinbei-shen committed:
Fix coding style violations

xinbei-shen committed:
Implement GUI using JavaFX

xinbei-shen committed:
Merge branch 'branch-Level-10'

xinbei-shen committed:
Add assertions

The code does not utilise any assertions.

Use assertions to verify assumptions about the program state.

Assertions help to indicate possible bugs in the code so that bugs and mistakes
can be more easily detected.

xinbei-shen committed:
Improve code quality and refactor code

Code quality is essential for readability and understandibility.

Add comments to explain certain aspects of the code and extract
methods such that SLAP is adhered to.

This would shorten long methods and make it easier for others
to understand the code.

xinbei-shen committed:
Merge pull request #2 from ichigh0st/branch-A-CodeQuality

Improve code quality and refactor code

xinbei-shen committed:
Merge branch 'branch-A-Assertions'

xinbei-shen committed:
Add tags to categorise tasks

Some tasks belong to similar categories.

Tagging tasks with tags can group the tasks for easier
searching of similar tasks later on.

xinbei-shen committed:
View tasks grouped under a specified tag

xinbei-shen committed:
Merge branch 'branch-C-Tagging'

xinbei-shen committed:
Refactor Parser and TaskList

xinbei-shen committed:
Include a user guide

xinbei-shen committed:
Update _config.yml
